test/unit/shmem_test.c

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (c) 2017 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+/* Each PE sends a message to every other PE.  PEs wait for all messages to
+ * arrive using shmem_test to poll the array. */
+
+#include <stdio.h>
+#include <shmem.h>
+
+/* Wait for any entry in the given ivar array to match the wait criteria and
+ * return the index of the entry that satisfied the test. */
+int wait_any(long *ivar, int count, shmem_cmp_t cmp, long value)
+{
+  int idx = 0;
+  while (!shmem_long_test(&ivar[idx], cmp, value))
+    idx = (idx + 1) % count;
+  return idx;
+}
+
+int main(void)
+{
+  shmem_init();
+  const int mype = shmem_my_pe();
+  const int npes = shmem_n_pes();
+
+  long *wait_vars = shmem_calloc(npes, sizeof(long));
+
+  /* Put mype+1 to every other PE */
```

Comment:
"every other PE" looks like "every PE" to me...  I have to be picky because this PR seems otherwise perfect.